staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/marshal_test.go

```diff
@@ -169,3 +432,105 @@ func TestJSONUnderlyingArrayReuse(t *testing.T) {
 		t.Errorf("unexpected mutation, got %s want %s", got, want)
 	}
 }
+
+func unmarshalJSONTest[T Testable[T]](input []byte, wantErr bool, expected T) func(t *testing.T) {
+	return func(t *testing.T) {
+		var actualDecoded T
+		actualDecoded = reflect.New(reflect.TypeOf(actualDecoded).Elem()).Interface().(T)
+		err := actualDecoded.UnmarshalJSON(input)
+		if (err != nil) != wantErr {
+			if wantErr {
+				t.Fatal("expected error")
+			}
+			t.Fatalf("unexpected error: %v", err)
+		}
+		if err == nil && !cmp.Equal(expected, actualDecoded) {
+			t.Error("unexpected decoded value")
+			t.Fatalf(cmp.Diff(expected, actualDecoded))
+		}
+	}
+}
+
+func roundTripJSONTest[T Testable[T]](input T, expectedEncoded []byte, expectedDecoded T) func(t *testing.T) {
+	return func(t *testing.T) {
+		input := input.DeepCopy()
+		actualEncoded, err := input.MarshalJSON()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if bytes.Compare(expectedEncoded, actualEncoded) != 0 {
```

Comment:
The linter thinks this should be `if !bytes.Equal(...)` but since the error text will include a diff, I prefer checking the string returned from cmp.Diff for emptiness as in my other suggestion.